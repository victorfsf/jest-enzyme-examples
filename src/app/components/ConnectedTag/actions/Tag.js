export const CHANGE_NAME = 'CHANGE_NAME';
export const TOGGLE_FORM = 'TOGGLE_FORM';

export const changeName = name => ({
  type: CHANGE_NAME,
  name,
});

export const toggleForm = () => ({
  type: TOGGLE_FORM,
});
