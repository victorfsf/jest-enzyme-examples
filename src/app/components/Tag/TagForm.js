import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { validateString } from 'app/utils';

import styles from './TagForm.scss';
import tagStyles from './Tag.scss';

class TagForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.name,
    };
    this.inputRef = React.createRef();
  }

  componentDidMount() {
    if (this.inputRef.current) {
      this.inputRef.current.focus();
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    const { onSubmit, onCancel } = this.props;
    const { value } = this.state;
    if (validateString(value)) {
      onSubmit(e, value);
      onCancel();
    }
  }

  handleChange(e) {
    const { value } = e.target;
    this.setState({ value });
  }

  render() {
    const { onCancel } = this.props;
    const { value } = this.state;
    return (
      <form
        className={classnames(tagStyles.tag, styles.tagForm)}
        onSubmit={e => this.handleSubmit(e)}
      >
        <input
          type="text"
          id="name"
          onChange={e => this.handleChange(e)}
          value={value}
          autoComplete="off"
          maxLength={35}
          placeholder="Please enter a tag name"
          ref={this.inputRef}
        />
        <button type="button" onClick={onCancel}>âœ–</button>
      </form>
    );
  }
}

TagForm.propTypes = {
  name: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default TagForm;
