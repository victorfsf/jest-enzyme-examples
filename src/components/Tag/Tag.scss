.tag {
  border: 1px solid #ddd;
  padding: 4px 9px;
  margin: 5px;
  border-radius: 2px;
  display: inline-block;
  font-size: 14px;
  background: #f6f6f6;
  color: #888;
  font-family: Lato, sans-serif;
  outline: 0;
}

.clickable {
  cursor: pointer;
}
