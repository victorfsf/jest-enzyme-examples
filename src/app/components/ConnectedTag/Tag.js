import React from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import classnames from 'classnames';

import { toggleForm } from './actions/Tag';
import TagForm from './TagForm';
import styles from './Tag.scss';

const Tag = ({ isFormOpen, name, toggleForm: handleToggleForm }) => (
  isFormOpen ? (
    <TagForm name={name} />
  ) : (
    <button
      className={classnames(styles.tag, styles.clickable)}
      onClick={handleToggleForm}
      type="button"
    >
      {name}
    </button>
  )
);

Tag.propTypes = {
  name: PropTypes.string.isRequired,
  isFormOpen: PropTypes.bool,
  toggleForm: PropTypes.func.isRequired,
};

Tag.defaultProps = {
  isFormOpen: false,
};

const mapStateToProps = state => state.tag; // state.tag = { name, isFormOpen }

const mapDispatchToProps = dispatch => bindActionCreators({
  toggleForm,
}, dispatch);


export default connect(mapStateToProps, mapDispatchToProps)(Tag);
