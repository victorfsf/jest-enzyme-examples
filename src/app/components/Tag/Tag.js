import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import TagForm from './TagForm';
import styles from './Tag.scss';

class Tag extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isFormOpen: false,
      name: props.name,
    };
  }

  toggleFormOpen() {
    const { isFormOpen } = this.state;
    this.setState({
      isFormOpen: !isFormOpen,
    });
  }

  handleSubmit(e, name) {
    this.setState({ name });
  }

  render() {
    const { isFormOpen, name } = this.state;
    const toggleForm = () => this.toggleFormOpen();
    return isFormOpen ? (
      <TagForm
        name={name}
        onCancel={toggleForm}
        onSubmit={(e, values) => this.handleSubmit(e, values)}
      />
    ) : (
      <button
        className={classnames(styles.tag, styles.clickable)}
        onClick={toggleForm}
        type="button"
      >
        {name}
      </button>
    );
  }
}

Tag.propTypes = {
  name: PropTypes.string.isRequired,
};

export default Tag;
