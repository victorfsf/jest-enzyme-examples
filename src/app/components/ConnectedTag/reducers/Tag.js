import { TOGGLE_FORM, CHANGE_NAME } from '../actions/Tag';

export default function tag(state = {}, action) {
  switch (action.type) {
    case CHANGE_NAME:
      return {
        ...state,
        name: action.name,
      };
    case TOGGLE_FORM:
      return {
        ...state,
        isFormOpen: !state.isFormOpen,
      };
    default:
      return state;
  }
}
