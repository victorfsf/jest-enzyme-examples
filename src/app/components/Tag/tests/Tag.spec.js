import React from 'react';
import { shallow } from 'enzyme';

import Tag from 'app/components/Tag';

const testTagChange = (wrapper, id, value) => {
  const tagForm = wrapper.dive();
  tagForm.find('input#name').simulate('change', { target: { id, value } });
  return tagForm;
};

describe('Tag (Enzyme only)', () => {
  test('renders a TagForm then cancels back to the tag', () => {
    const wrapper = shallow(<Tag name="test" />);
    expect(wrapper.find('button.tag.clickable')).toHaveLength(1);
    expect(wrapper.find('TagForm')).toHaveLength(0);

    wrapper.simulate('click');
    expect(wrapper.find('button.tag.clickable')).toHaveLength(0);
    expect(wrapper.find('TagForm')).toHaveLength(1);

    wrapper.dive().find('button[type="button"]').at(0).simulate('click');
    expect(wrapper.find('button.tag.clickable')).toHaveLength(1);
    expect(wrapper.find('TagForm')).toHaveLength(0);
  });

  test('changes the tag name', () => {
    const wrapper = shallow(<Tag name="test" />);
    wrapper.simulate('click');

    const tagForm = testTagChange(wrapper, 'name', 'this is a new name');
    tagForm.simulate('submit', {
      preventDefault: jest.fn(),
    });

    expect(wrapper.state('name')).toEqual('this is a new name');
  });

  test('doesnt change the tag name if cancel is clicked', () => {
    const wrapper = shallow(<Tag name="test" />);
    wrapper.simulate('click');

    const tagForm = testTagChange(
      wrapper,
      'name',
      'this was supposed to be a new name',
    );
    tagForm.find('button[type="button"]').at(0).simulate('click');

    expect(wrapper.state('name')).toEqual('test');
  });

  test('doesnt change the tag name if name.length === 0', () => {
    const wrapper = shallow(<Tag name="test" />);
    wrapper.simulate('click');

    const tagForm = testTagChange(wrapper, 'name', '');
    tagForm.simulate('submit', {
      preventDefault: jest.fn(),
    });

    expect(wrapper.state('name')).toEqual('test');
  });
});

describe('Tag (Snapshots)', () => {
  test('renders a tag', () => {
    const wrapper = shallow(<Tag name="test" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders a TagForm when the tag is clicked', () => {
    const wrapper = shallow(<Tag name="test" />);
    wrapper.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  test('renders a tag with a changed name', () => {
    const wrapper = shallow(<Tag name="test" />);
    wrapper.simulate('click');

    const tagForm = testTagChange(wrapper, 'name', 'this is a new name');
    tagForm.simulate('submit', {
      preventDefault: jest.fn(),
    });

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Tag (Mocks)', () => {
  test('calls toggleFormOpen when the tag is clicked', () => {
    const wrapper = shallow(<Tag name="test" />);
    const mockedFormOpen = jest.fn();
    wrapper.instance().toggleFormOpen = mockedFormOpen;
    wrapper.simulate('click');
    expect(mockedFormOpen).toHaveBeenCalled();
  });
});
