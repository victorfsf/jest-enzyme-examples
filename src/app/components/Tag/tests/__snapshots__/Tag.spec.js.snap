// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag (Snapshots) renders a TagForm when the tag is clicked 1`] = `
<TagForm
  name="test"
  onCancel={[Function]}
  onSubmit={[Function]}
/>
`;

exports[`Tag (Snapshots) renders a tag 1`] = `
<button
  className="tag clickable"
  onClick={[Function]}
  type="button"
>
  test
</button>
`;

exports[`Tag (Snapshots) renders a tag with a changed name 1`] = `
<button
  className="tag clickable"
  onClick={[Function]}
  type="button"
>
  this is a new name
</button>
`;
